### ISSUE TITLE
HostBinding can't bind to host input property other than `class.`, `attr.` and `style.`

### ISSUE THREAD

**Issue by maximusk** [angular/angular#16797]
**I'm submitting a ...**  (check one with "x")
```
[x ] bug report
```

**Current behavior**
When binding to a custom input property `acproperty` on the host component using `@HostBinding()` I get an error:

> Can't bind to 'acproperty' since it isn't a known property of 'a-comp'

**Expected behavior**
I expect `@HostBinding()` decorator to bind to any property on the host component, not just the ones with `class.`, `attr.` and `style.`

**Minimal reproduction of the problem with instructions**
[Here is the Plunker](http://plnkr.co/edit/b9zFc9ZqxfYHz5o0xy4t?p=preview)

Windows10, Angular 4.x, Chrome

**Comment by mlc-mlapis** [comment]
Probably very unreadable finally.

**Comment by maximusk** [comment]
@mlc-mlapis, what do you mean?

**Comment by mlc-mlapis** [comment]
I meant how that could affect the behaviour of the parent component if it would be implemented and the readability of such relations. Because if you want to do such things you can just inject the component through DI and make the same operations.

**Comment by maximusk** [comment]
@mlc-mlapis ,  I don't really understand your point. This seems to be a bug, because `@HostBinding('style.backgroundColor') c_colorrr = "red";` is working fine. And as I understand this is the purpose of `HostBinding` - to update property on a host element be it an HTML element or a component.

**Comment by trotyl** [comment]
@mlc-mlapis 

> If a binding changes, it will update the **host element** of the directive.

> `HostBinding` takes an optional parameter that specifies the property name of the **host element** that will be updated.

I don't think there were documentation saying that `@HostBinding()` can be used to bind Component property, could you provide that?

See [HostBindingDecorator](https://github.com/angular/angular/blob/71f5b73296708014b740fb5dd0145c0599de7a19/packages/core/src/metadata/directives.ts#L902-L938) for more information.

**Comment by trotyl** [comment]
@maximusk I think you may have some misunderstanding, `@HostBinding()` always binding to the DOM element, you can change your code to 

```typescript
@HostBinding('id') my = 'some';
```

You will find it's working fine.

**Comment by maximusk** [comment]
> I don't think there were documentation saying that @HostBinding() can be used to bind Component property, could you provide that?
I don't think there is a documentation for `@HostBinding()` at all.

> @HostBinding() always binding to the DOM element
What makes you think so? My understand of the way Angular was designed is to remove a distinction between DOM elements and Components. For example, the syntax for binding is the same. Besides, `@HostListener()` can listen to both DOM element and Components events.

**Comment by trotyl** [comment]
What you think is not possible for one-way data flow.

**Comment by mlc-mlapis** [comment]
@maximusk ... there is the known issue for this problem and actually it is marked as a bug https://github.com/angular/angular/issues/13776 (https://github.com/angular/angular/issues/10499). I thought about it once more and it is true that it should works as you described. It's nothing against anything I guess. But probably the problem has not so high priority actually.

**Comment by trotyl** [comment]
@mlc-mlapis Thanks for the information, seems they do mark it as BUG after enough thumbs up. But I don't think that's fixable, just like https://github.com/angular/angular/issues/7329.

**Comment by maximusk** [comment]
@mlc-mlapis , thanks for the link, I'll check it.

> What you think is not possible for one-way data flow.
If you view a directive class instance as a child of a component, then it indeed goes against one-way data flow. But I see a directive as an extension of a component, not its child.

**Comment by trotyl** [comment]
@maximusk 
`@Input` not only applies to Component, but also (other) Directive.
Consider ADirective has an `@Input()` of property `a` and an `@HostBinding()` of property `b` with property `a` being a setter to update property `b`, while BDirective has an `@Input()` of property `b` and an `@HostBinding()` of property `a` with property `b` being a setter to update property `a`, what should be then?

**Comment by mlc-mlapis** [comment]
@trotyl ... thanks for your example. Yes, it is true and it is really a question. But there are other similar cases as changings done during lifecycle hooks that lead to the same situation ... throw change-after-checked errors in dev mode.

So it won't be an extra exception but still it will be other additional place with potential problems ...

**Comment by trotyl** [comment]
Indeed it could just throw exceptions, but since ADirective & BDirective are both sound Directives with valid one-way data flow, but combining them together causing an invalid two-way data flow, that't would be a design failure.

**Comment by mlc-mlapis** [comment]
@trotyl ... I agree that the idea of supporting that would be a counterproductive step and I also think that Angular core team won't accept it in the end.

**Comment by trotyl** [comment]
https://github.com/angular/angular/issues/16493 may also be related to this issue.

**Comment by trotyl** [comment]
https://github.com/angular/angular/issues/16493 may also be related to this issue in:

> @HostBinding is always directly to the DOM.

Not sure about their discussion context.

**Comment by mlc-mlapis** [comment]
Yes, all the discussions leads to that conclusion.

**Comment by tbosch** [comment]
This works as intended, as:
- using data binding to communicate between directives / components on the same element is slower than direct communication by making one inject the other
- data binding between directives easily leads to cycles.

You are right though in that `@HostListener` does work for events fired by directives as well, and that this fact is confusing.

Closing in favor of #16797.

**Comment by maximusk** [comment]
@tbosch, thanks, can you please provide and example for

> using data binding to communicate between directives / components on the same element is slower than direct communication **by making one inject the other**

 do you mean inject  using `@ViewChild`?

**Comment by trotyl** [comment]
@maximusk Just use the normal DI way in constructor as if it's a service, an example can be found in: http://plnkr.co/edit/BxCmIX1Ghkx39WgdzJAv?p=preview

**Comment by mlc-mlapis** [comment]
And also opposite as
```
@Directive({
  selector: '[some-directive]'
})
export class SomeDirective {
  constructor(private someComponent: SomeComponent)
}
```

**Comment by maximusk** [comment]
@trotyl, @mlc-mlapis  thank you guys for your suggestions.

**Comment by mlc-mlapis** [comment]
@maximusk You are right, it is not necessary and this should also fine:

`providers: [{ provide: Parent, useClass: AnyComponent }]`

Right now I am not sure why I remember this with `forwardRef`, maybe some more complicated case.

**Comment by trotyl** [comment]
In JavaScript, `class` does not have hoisting behavior, unlike `function`. So when you're using a class that being wrote below the using position, it's value would be undefined at that time.

Then `forwardRef` would come to help, it would postpone the evaluation time so Angular can get that initialized value.

**Comment by mlc-mlapis** [comment]
@trotyl Ah, thanks for your notice. Actually when using 1:1 = TS file : component definition, it seems to be unnecessary, but there are some cases that it could really help.

**Comment by zlepper** [comment]
Sorry to necro a semi old issue, but i think there might actually be something about this. 

If you want to access any property that is part of the HTMLElement, but not on either attr, class or style, then you have to fallback to using elementRef. This is not satisfactory when attempting to run in a web-worker, as any chance of doing that goes out the window, then moment the property has to be accessed. 

You would imagine something like this would work:
```ts

@Directive({
    selector: '[child]'
})
export class ChildComponent {

    @HostBinding('offsetHeight')
    public height: number;

}

```

**Comment by trotyl** [comment]
@zlepper Is it related to https://github.com/angular/angular/issues/8785 or https://github.com/angular/angular/issues/13776?

**Comment by zlepper** [comment]
I don't believe so.

They are talking about using one component/directive to affect other components/directives (at least as I read them).

Where as I'm talking about affecting/reading from the native component, something that you right now has to do using the native element. (elementRef)

**Comment by trotyl** [comment]
@zlepper From [MDN](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/offsetHeight):

> The HTMLElement.offsetHeight **read-only property** is the height of the element including vertical padding and borders, as an integer.

Of course you cannot set value to it. You can try using something like `className` and would found it's working fine.

**Comment by zlepper** [comment]
Yes, and I'm indeed interested in reading from the property.
I can't use style and classname, as I want to work with something that has dynamic heights, depending on the content. 


The reason I mentioned writing to the property, is I think it would be very inline with also adding read capabilities. But I would be just fine with reading for my actual usecase.

**Comment by trotyl** [comment]
@zlepper Since `@HostBinding()` is only meant for setting value, this is currently the expected behavior, I think you can make a new proposal about mechanism for reading DOM properties.

**Comment by zlepper** [comment]
Alright, will do. Thank you :)

**Comment by danielrsantana** [comment]
I know this question is quite old but I'm facing the same problem as @zlepper and I don't know the best way to achieve/workaround what I'm trying to do:

I have a component, raptor-button that receive css classes from @Input() cssClass. As I was not aware that angular directives were related only to DOM attributes, I've developed some directives to easy the path to some bootstrap classes, like bellow:

`@Directive({
  selector: '[raptor-btn-danger]'
})
export class RaptorDangerButtonCssDirective {

  @HostBinding('cssClass') classes = 'btn btn-danger';
}`

and to put that in use I've included my new directives in module file and consumed them in my page component, planning to replace the repetitive class declarations (they are bigger), like bellow:

**instead of:**
`<raptor-button cssClass="btn btn-sm btn-outline-primary btn-footer" text="Page View" icon="fa fa-clone"></raptor-button>`

**i've planned**
`<raptor-button **raptor-btn-primary-outline-sm** text="Page View" icon="fa fa-clone"></raptor-button>`

But then ng build said: "_Can't bind to 'cssClass' since it isn't a known property of 'raptor-button'_".

Any suggestions on how to achieve what I've planned for directives? I've tried also to change the Input() name to "class" but by doing that, the component itself received the class but the component property remained empty.

Thank you.
